api/src/neo4j/queries/map.js

```diff
@@ -0,0 +1,75 @@
+import querySingleResult from 'neo4j/queryHandlers/single';
+import queryListResult from 'neo4j/queryHandlers/list';
+
+const getMapsListing = async ({ model, version }) => {
+  const m = model;
+  const v = version;
+
+  const statement = `
+MATCH (:CompartmentState)-[:V${v}]-(c:Compartment:${m})
+CALL apoc.cypher.run("
+  MATCH (cs:CompartmentState)-[:V${v}]-(:Compartment:${m} {id: $cid})
+  RETURN cs { id: $cid, .* } as data
+  
+  UNION
+  
+  MATCH (:Compartment:${m} {id: $cid})-[:V${v}]-(:CompartmentalizedMetabolite)-[:V${v}]-(r:Reaction)
+  RETURN { id: $cid, reactionCount: COUNT(DISTINCT(r)) } as data
+  
+  UNION
+  
+  MATCH (csvg:SvgMap)-[:V${v}]-(:Compartment:${m} {id: $cid})
+  RETURN { id: $cid, compartmentSVGs: COLLECT(DISTINCT(csvg {.*})) } as data
+", {cid:c.id}) yield value
+RETURN { compartment: apoc.map.mergeList(apoc.coll.flatten(
+	apoc.map.values(apoc.map.groupByMulti(COLLECT(value.data), "id"), [value.data.id])
+)) } as data
+
+UNION
+
+MATCH (:SubsystemState)-[:V${v}]-(s:Subsystem:${m})
+CALL apoc.cypher.run("
+  MATCH (ss:SubsystemState)-[:V${v}]-(s:Subsystem {id: $sid})
+  RETURN ss { id: $sid, .* } as data
+  
+  UNION
+  
+  MATCH (:Subsystem:${m} {id: $sid})-[:V${v}]-(r:Reaction)
+  RETURN { id: $sid, reactionCount: COUNT(DISTINCT(r)) } as data
+  
+  UNION
+  
+  MATCH (:Subsystem:${m} {id: $sid})-[:V${v}]-(ssvg:SvgMap)
+  RETURN { id: $sid, subsystemSVGs: COLLECT(DISTINCT(ssvg {.*})) } as data
+", {sid:s.id}) yield value
+RETURN { subsystem: apoc.map.mergeList(apoc.coll.flatten(
+	apoc.map.values(apoc.map.groupByMulti(COLLECT(value.data), "id"), [value.data.id])
+)) } as data
+`;
+
+  const result = await queryListResult(statement);
+
+  const mapListing = {
+    compartments: result.filter(o => !!o.compartment).reduce((l, o) => [...l, o.compartment] , []),
+    subsystems: result.filter(o => !!o.subsystem).reduce((l, o) => [...l, o.subsystem] , []),
+  };
+
+  return mapListing;
+};
+
+const mapSearch = async ({ model, version, searchTerm }) => {
+  const statement = `
+CALL db.index.fulltext.queryNodes("fulltext", "${searchTerm}~")
+YIELD node
+OPTIONAL MATCH (node)-[:V${version}]-(parentNode:${model})
+WHERE node:${model} OR parentNode:${model}
+RETURN [n in COLLECT(DISTINCT(
+	CASE
+		WHEN EXISTS(node.id) THEN node.id
+		ELSE parentNode.id
+	END
+))[..100] WHERE n IS NOT NULL]`;
```

Comment:
This is the current search algorithm: https://github.com/MetabolicAtlas/MetabolicAtlas/blob/master/backend/api/private_views.py#L105 I thought it was similar enough to the current gem search, which is also not ported over, and was unsure how long it would take to investigate and fully replicate. So I thought it makes sense to mention it for now but create a separate issue for.

Reply:
The Django search was exact (case insensitive), so more restrictive than the gem search. I would also suggest to stick to i-exact search here, because there is no listing of the results returned allowing to easily filter the matches. The matching SVG nodes get highlighted and with this search algorithm you might end up to HL to many of them, confusing the users.

I would add also that the two search algorithms, map search and gem / global search were not so different when using Django, But if you decide to set this one on exact search only, users might become frustrated by the different behaviors.

Also we don't always expect to get exactly one match, e.g. the user searches for 'ATP', all the different compartmentalizedMetabolite nodes should be returned. Same if the user searches a metabolite formula or an EC code to find reactions.

Reply:
Thanks for the input @pecholleyc. I see that most of the queries are `*_iexact` ones but there is still a few `*_iregex` ones, but maybe those are still kind of exact as well with some characters being replaced/omitted. I also saw there were some lines like `APImodels.ReactionComponent.objects.using(model).prefetch_related('external_databases')...` that I'm not familiar with either, which made it difficult for me to give a good estimation of this task without taking more time to understand the current implementation.

Do you know roughly how long it would take to reimplement map search in JavaScript and Cypher? I still would like to suggest that if this would take more than several hours, it is still worthwhile to see how the current search works with the map. If we always get good enough results then maybe we can leave it as is.

Reply:
Let's go with the last suggestion from @e0 - this is marked as done, and we have MA-887 to follow up on the behaviour of the search.

Reply:
Let's go with the last suggestion from @e0 - this is marked as done, and we have MA-887 as a follow-up on the search behaviour.